[FEAT] 태그 라벨링 API 구조 개선 및 LLM 서비스 분리

Callein opened the pull request:
## 주요 변경사항

1. **파일명 수정**  
   - `tag_labeling_api.py`를 `tag_labeling_router.py`로 변경하여 명칭 일관성을 유지했습니다.

2. **requirements.txt 업데이트**  
   - Google Gemini API(google-genai) 사용을 위한 패키지를 추가하고, 불필요한 항목을 정리했습니다.

3. **환경변수 추가 (`EnvVariables`)**  
   - `GEMINI_API_KEY`, `GEMINI_API_URL` 추가
   - `LLM_PROVIDER`설정을 통해 Ollama와 Gemini 중 선택적으로 사용할 수 있게 했습니다.

4. **LLM 로직 분리 및 `request_tag_assignment` 구현 (LLMService)**  
   - LLM 관련 메서드를 별도의 `LLMService` 클래스로 분리하여 관리했습니다.  
   - `LLM_PROVIDER` 값을 확인해 Ollama 또는 Gemini API를 호출하도록 `request_tag_assignment` 메서드를 구현했습니다.

5. **메서드 문서화 (docs):**  
   - `assign_tags_to_messages_iterative`에 대해 종합적인 docstring을 추가, 함수의 동작과 인자/반환값을 명시했습니다.

## 개선사항
- **태그 라벨링 API**와 **LLM 호출 로직**이 명확히 분리
- Ollama와 Gemini API를 환경변수 설정에 따라 손쉽게 전환 가능
- `gemini-2.0-flash-lite` 사용으로, 응답속도 약 **60%** 개선 및 정확도 향상

  <details>
  <summary>PPI masking 된 input messages</summary> 
  
    ```json
    [
      {
        "subject_id": "1621",
        "masked_message": "아이폰 13 미니 128GB 블루 판매합니다! 📱\n가격: 40만원\n구성: 풀박스 + 케이스 3개 + 보호 필름 부착\n충전 케이블 포함 (가끔 작동 안 될 때 있음)\n기기 상태: 생활 기스 있음, 배터리 최대 용량 80%\n궁금한 점 있으시면 오픈채팅으로 들어와서 편하게 물어봐 주세요 😊\n[링크]"
      },
      {
        "subject_id": "1629",
        "masked_message": "🍀경영경제학부 임원단 모집🍀\n\n안녕하세요. 경영경제학부 여러분\n25-1 학기 동안 경영경제학부를 함께 이끌어 갈 임원단을 모집합니다!\n경영경제학부 임원단은 학부를 섬기며 각종 행사를 돕고 진행하는 자리입니다.\n경경인을 위하는 마음으로 학부생들을 섬기고 대표단을 도와 경영경제학부를 더 나은 학부로 이끌어 나갈 경경 임원단을 모집합니다.\n\n✅모집 기한: 3/15(토) 12:00 ~ 3/25(화) 23:59\n\n✅모집 분야\n📌총무 (1명)\n- 학부 운영비를 관리합니다. 열정과 노력하는 자세가 필요합니다. 대표가 많이 도와드립니다.\n📌서기(1명)\n- 임원단 회의 시에 회의록을 작성합니다. 경청하는 자세와 타자 칠 손이 필요합니다.\n📌디자이너 (1명)\n- 각종 행사 포스터 등을 제작합니다. 미적인 감각과 열정이 필요합니다.\n📌일반 임원 (0명)\n- 각종 학부 행사 시에 대표단과 다른 임원들을 도와 행사 보조 및 진행 역할을 합니다. 성실한 자세와 열정, 학부를 사랑하는 마음이 필요합니다.\n\n✅모집자격: 한동대학교 재학생\n\n⁉️신청방법: 구글폼 작성\n[링크]\n\n많은 관심과 지원 부탁드립니다!!❤️\n\n문의)\n[전화번호] 대표 [이름]\n[이메일]"
      },
      {
        "subject_id": "1622",
        "masked_message": "🔥 한동BST(돌파구전략훈련)학교에 여러분을 초대합니다! 🔥\n\n[이름] 그리스도의 사랑을 알고(💙), 누리고(🙌), 나누고(🤝), 전하는(📢) 법을 배우고 훈련합니다 \n\n✨ 왜 이 훈련을 받아야 할까요?\n✅ 복음에 대한 정확한 이해   \n✅ 복음 전파의 구체적 방법과 실천  \n✅ 말씀과 기도와 순종의 삶\n✅ 함께 이루는 하나님 나라 \n\n🚀 이 훈련을 통해 얻게 될 변화\n📖 말씀과 기도의 삶이 자리 잡힙니다\n🗣 복음을 명확하게 이해하고, 담대하게 전합니다  \n🔥 성령님과 동행하며 순종으로 능력과 열매를 경험합니다\n🎯 하나님 나라를 함께 이루어 갈 멘토와 동역자를 만납니다\n❤️ 한동을 시작한 1회 졸업생 선배들이 직접 가르치고 훈련합니다 \n\n📌 훈련 일정 안내\n📅 일정: 학기 중 월요일 오후 7시~9시 (4주차 3/24 시작, 중간·기말고사 기간 제외, 총 9주)  \n📍 장소: 학관 101B  \n👥 인원: 30명 (선착순 마감)  \n💰 비용: 무료  \n🔗 신청 링크:[링크]"
      },
      {
        "subject_id": "1633",
        "masked_message": "🌿영국의 작은 전환마을 토트네스 다큐 전시 초대 🌿\n\n여러분, 전환마을이라는 개념 들어보셨나요?\n기후위기, 자원 고갈 속에서도 지속가능한 삶을 실천하는 마을, 영국 토트네스.\n우리는 그곳을 직접 찾아가 변화의 모습을 기록했고, 이 작은 마을에서 세상을 어떻게 바꾸고 있는지 여러분들께 전달드리고자합니다.\n\n📅 전시 일정: 3.18(화)~ 3.19(수)\n📍 장소: 학관 2층 중구난방\n\n🍿 다큐 상영 일정: 3.18(화)~3.19(수): \n      1타임: 19:00 / 2타임: 21:00\n📍 장소: 18일(화) 학관 101B\n               19일(수) 학관 101A\n\n지속가능한 미래에 대해 고민하고, 새로운 가능성을 함께 발견하고 싶다면❓\n환경을 생각하는 여러분들을, 토트네스 마을로 초대합니다 💌\n\n궁금한 점이 있다면 편하게 문의 주세요 :)\n연락처: 누구꺼 적을까요?\n많은 관심과 방문 기다리겠습니다 🌱"
      },
      {
        "subject_id": "1631",
        "masked_message": "☘️안녕하세요, 저희는 졸업연구 프로젝트를 진행하고 있는 [이름]의 [이름], [이름], [이름]입니다 :)🐹\n\n현재 저희는 Display 관련 프로젝트를 진행하고 있으며, \n사전에 저희가 진행한 리서치 및 평가를 바탕으로 HOME Display를 활용하고 있는 다양한 🔥행태 파악🔥을 위해 설문 조사를 진행하고자 합니다. 🖥️\n\n본 설문 조사는 사용자들이 어떤 식으로 집에서 Display를 사용하는지를 알아보기 위한 가벼운 목적이니 부담가지지 않으셔도 됩니다.🙃\n\n예상 소요 시간은 약 3-5분 입니다.\n\n설문 참여 시 추첨을 통해 소정의 기프티콘을 드릴 예정입니다.✨\n\n💁‍♂️ 참여 대상: 디스플레이를 사용하는 GenZ세대\n🗓️ 설문조사 기간: 3월 18일 (화) ~ 3월 20일 (목)\n\n[링크]\n\n▪ 문의: [전화번호]"
      },
      {
        "subject_id": "1636",
        "masked_message": "😱 분실물을 찾습니다!! 😱\n\n오석관 3층 5열람실 앞 책장에 뒀던 민법총칙, 성격심리학 책을 분실했습니다…! 🥺\n지난주 주말(3/14~16) 본가에 다녀온 사이에 없어진 것 같습니다ㅠ\n혹시 실수로 가져가셨다면 원래 있던 곳에 돌려주시면 감사하겠습니다……… \n보신 적 있다면 아래 번호로 연락주세요! 정말 소중한 책들입니다……  😭😱\n[전화번호]"
      },
      {
        "subject_id": "1637",
        "masked_message": "[Yesod 조각피자 할인 Event]\n\n🎁Yesod 조각피자 할인🎁\n\n🧀치즈 3900 → 3000\n🍕페퍼로니 4200 → 3000\n🍍하와이안 4200 → 3000\n🍗핫치킨 4600 → 3000\n\n📢 해당 이벤트는 선착순 마감으로 진행됩니다! 따라서 조기 마감될 수 있음을 미리 알려드립니다!📢\n\n문의사항 : [링크]"
      },
      {
        "subject_id": "1634",
        "masked_message": "🤎2025년 세워가는 세대 새섬 대상 또래상담 멘토링 프로그램 멘토 모집🤎\n\n💬: 우와, 새내기들의 멘토가 될 뿐만 아니라 활동비 지원까지 받을 수 있는 프로그램이있다고❓\n\n한동대학교 상담센터에서 또래상담 멘토링 프로그램에 활동할 세워가는 세대 새내기 섬김이들을 모집합니다‼️\n“또래상담 멘토링 프로그램”은 새내기 섬김이에게 준전문상담 훈련의 기회를 제공하여 담당 새내기들에게 심리정서적인 도움을 효과적으로 줄 수 있도록 돕는 대학생활 적응 지원 프로그램입니다."
      }
    ]
  
    ```
  </details>

  <details>
  <summary>`llama3:8b` 답변 (15.80s)</summary>
  
  ```json
  {
    "assignments": [
      {
        "subject_id": "1621",
        "tag_codes": [
          "sale",
          "recruit",
          "sharing"
        ]
      },
      {
        "subject_id": "1629",
        "tag_codes": [
          "recruit",
          "promotion"
        ]
      },
      {
        "subject_id": "1622",
        "tag_codes": [
          "recruit",
          "performance"
        ]
      },
      {
        "subject_id": "1633",
        "tag_codes": [
          "event",
          "promotion"
        ]
      },
      {
        "subject_id": "1631",
        "tag_codes": [
          "event",
          "survey"
        ]
      },
      {
        "subject_id": "1636",
        "tag_codes": [
          "lost-found",
          "sharing"
        ]
      },
      {
        "subject_id": "1637",
        "tag_codes": [
          "event",
          "promotion"
        ]
      },
      {
        "subject_id": "1634",
        "tag_codes": [
          "promotion",
          "recruit"
        ]
      }
    ]
  }
  ```
  </details>
  
  <details>
  <summary>`gemini-2.0-flash-lite` 답변 (6.36s)</summary>
     
  ```json
  {
    "assignments": [
      {
        "subject_id": "1621",
        "tag_codes": [
          "sale"
        ]
      },
      {
        "subject_id": "1629",
        "tag_codes": [
          "recruit"
        ]
      },
      {
        "subject_id": "1622",
        "tag_codes": [
          "event",
          "faith",
          "recruit"
        ]
      },
      {
        "subject_id": "1633",
        "tag_codes": [
          "event",
          "promotion"
        ]
      },
      {
        "subject_id": "1631",
        "tag_codes": [
          "survey",
          "promotion"
        ]
      },
      {
        "subject_id": "1636",
        "tag_codes": [
          "lost-found"
        ]
      },
      {
        "subject_id": "1637",
        "tag_codes": [
          "event",
          "promotion",
          "sale"
        ]
      },
      {
        "subject_id": "1634",
        "tag_codes": [
          "recruit"
        ]
      }
    ]
  }
  ```
  </details>
  
## 관련 이슈
resolves #6 
resolves #10 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **새로운 기능**
	- 메시지 태그 할당 프로세스가 강화되어, 첨단 언어 모델 통합을 통해 다양한 공급자 옵션(예: Gemini, Ollama)으로 더욱 안정적이고 일관된 태그 지정이 가능해졌습니다.
	- 환경 변수 템플릿 파일 `.env.example`가 추가되어, 사용자 설정을 위한 다양한 변수들이 정의되었습니다.
- **잡일**
	- 환경 구성을 위해 새 변수들이 도입되어, 관련 설정 옵션이 확장되었습니다.
	- 서비스 안정성과 기능 강화를 위해 새로운 라이브러리 의존성이 추가되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Callein committed:
:pencil2: chore: rename typo in tag_labeling_api.py to tag_labeling_router.py

Callein committed:
📦 Update requirements.txt - google-genai

Callein committed:
🔐 chore: add GEMINI_API_KEY, GEMINI_API_URL, LLM_PROVIDER to EnvVariables

Callein committed:
✨ feat: Separate LLM methods to LLMService and implement request_tag_assignment using Gemini and Ollama
  - Extract LLM logic into a dedicated LLMService that selects between Gemini and Ollama API calls based on EnvVariables.LLM_PROVIDER

Callein committed:
:memo: docs: add comprehensive docstring to assign_tags_to_messages_iterative method

- Provide detailed explanation of the method workflow, its arguments, and the return value.
- The docstring now clearly describes the masking, cleaning, LLM call, and assembly of tag assignments.

Callein committed:
:recycle: refactor: apply LLMService.assign_tag_to_message

junglesub reviewed (approved):
잘 작동된다니 다행입니다!

junglesub commented:
> 분당 30 요청
> 분당 100만 토큰
> 일당 1500 요청
> https://github.com/handong-app/handong-feed-spotlight/issues/10#issuecomment-2798891821

일당 1500 요청은 spotlight 만 하는 것은 문제가 없어 보이네요. 다만 분당 30 요청을 준수하기 위해 RateLimit 장치가 준비되어 있을까요?

Callein commented:
아직 하루에 30 요청이 들어오는 경우는 드물어서, Priority 1로 다음 주차에 진행해도 괜찮을까요? 
일단 external API 연결이 더 우선순위 높을 것 같네요!

코드레빗 nitpick 커멘트만 해결하고 머지하겠습니다!

Callein committed:
:memo: docs: Fix request_tag_assignment document - 올라마 to Ollama

junglesub commented:
네 알겠습니다. 머지해주시고 api 연동 테스트 후에 점진적으로 적용해보져!

Callein committed:
:pencil2: chore: remove useless semicolon

Callein committed:
:memo: docs: add .env.example

Callein committed:
:white_check_mark: chore: add log when no valid assignment returned for subject_id

Callein committed:
:memo: docs: update assign_tag_to_message log to show accurate llm provider

Callein committed:
🔐 chore: add OLLAMA_MODEL to EnvVariables

Callein committed:
:white_check_mark: apply EnvVariables.OLLAMA_MODEL to ollama.chat(model=)

Callein committed:
📝 docs: update assign_tag_to_message docs

Callein committed:
🐛 fix: add environment variable check for OLLAMA_MODEL

- Raise an exception if EnvVariables.OLLAMA_MODEL is not set.

Callein committed:
✨ docs: add environment variable explanations to .env.example
  - Provide descriptions for API_PORT, DB settings, BASE_URL, and LLM_PROVIDER
  - Include explanations for HUGGING_FACE_TOKEN, OLLAMA_MODEL, GEMINI_API_KEY, GEMINI_API_URL, and test data variables SAMPLE_TAGS/SAMPLE_MESSAGES
